package com.dwarfeng.essentials.node.minimal.he.configuration;

import com.dwarfeng.acckeeper.sdk.bean.BeanMapper;
import com.dwarfeng.acckeeper.sdk.bean.entity.*;
import com.dwarfeng.acckeeper.sdk.bean.key.formatter.ProtectorVariableStringKeyFormatter;
import com.dwarfeng.acckeeper.sdk.bean.key.formatter.RecordStringKeyFormatter;
import com.dwarfeng.acckeeper.stack.bean.entity.*;
import com.dwarfeng.acckeeper.stack.bean.key.ProtectorVariableKey;
import com.dwarfeng.acckeeper.stack.bean.key.RecordKey;
import com.dwarfeng.subgrade.impl.bean.MapStructBeanTransformer;
import com.dwarfeng.subgrade.impl.cache.RedisBatchBaseCache;
import com.dwarfeng.subgrade.sdk.redis.formatter.LongIdStringKeyFormatter;
import com.dwarfeng.subgrade.sdk.redis.formatter.StringIdStringKeyFormatter;
import com.dwarfeng.subgrade.stack.bean.key.LongIdKey;
import com.dwarfeng.subgrade.stack.bean.key.StringIdKey;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.core.RedisTemplate;

@Configuration
public class AcckeeperCacheConfiguration {

    private final RedisTemplate<String, ?> template;

    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.account}")
    private String accountPrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.login_state}")
    private String loginStatePrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.login_history}")
    private String loginHistoryPrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.protector_info}")
    private String protectorInfoPrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.protector_support}")
    private String protectorSupportPrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.protector_variable}")
    private String protectorVariablePrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.login_param_record}")
    private String loginParamRecordPrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.protect_detail_record}")
    private String protectDetailRecordPrefix;
    @Value("${com.dwarfeng.essentials.cache.prefix.entity.acckeeper.derive_history}")
    private String deriveHistoryPrefix;

    public AcckeeperCacheConfiguration(RedisTemplate<String, ?> template) {
        this.template = template;
    }

    @Bean(name = "acckeeper.accountCacheDelegate")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<StringIdKey, Account, FastJsonAccount> accountCacheDelegate() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonAccount>) template,
                new StringIdStringKeyFormatter(accountPrefix),
                new MapStructBeanTransformer<>(Account.class, FastJsonAccount.class, BeanMapper.class)
        );
    }

    @Bean(name = "acckeeper.loginStateRedisBatchBaseCache")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<StringIdKey, LoginState, FastJsonLoginState> loginStateRedisBatchBaseCache() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonLoginState>) template,
                new StringIdStringKeyFormatter(loginStatePrefix),
                new MapStructBeanTransformer<>(LoginState.class, FastJsonLoginState.class, BeanMapper.class)
        );
    }

    @Bean(name = "acckeeper.loginHistoryCacheDelegate")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<LongIdKey, LoginHistory, FastJsonLoginHistory> loginHistoryCacheDelegate() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonLoginHistory>) template,
                new LongIdStringKeyFormatter(loginHistoryPrefix),
                new MapStructBeanTransformer<>(LoginHistory.class, FastJsonLoginHistory.class, BeanMapper.class)
        );
    }

    @Bean(name = "acckeeper.protectorInfoRedisBatchBaseCache")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<StringIdKey, ProtectorInfo, FastJsonProtectorInfo>
    protectorInfoRedisBatchBaseCache() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonProtectorInfo>) template,
                new StringIdStringKeyFormatter(protectorInfoPrefix),
                new MapStructBeanTransformer<>(ProtectorInfo.class, FastJsonProtectorInfo.class, BeanMapper.class)
        );
    }

    @Bean(name = "acckeeper.protectorSupportRedisBatchBaseCache")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<StringIdKey, ProtectorSupport, FastJsonProtectorSupport>
    protectorSupportRedisBatchBaseCache() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonProtectorSupport>) template,
                new StringIdStringKeyFormatter(protectorSupportPrefix),
                new MapStructBeanTransformer<>(
                        ProtectorSupport.class, FastJsonProtectorSupport.class, BeanMapper.class
                )
        );
    }

    @Bean(name = "acckeeper.protectorVariableRedisBatchBaseCache")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<ProtectorVariableKey, ProtectorVariable, FastJsonProtectorVariable>
    protectorVariableRedisBatchBaseCache() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonProtectorVariable>) template,
                new ProtectorVariableStringKeyFormatter(protectorVariablePrefix),
                new MapStructBeanTransformer<>(
                        ProtectorVariable.class, FastJsonProtectorVariable.class, BeanMapper.class
                )
        );
    }

    @Bean(name = "acckeeper.loginParamRecordCacheDelegate")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<RecordKey, LoginParamRecord, FastJsonLoginParamRecord>
    loginParamRecordCacheDelegate() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonLoginParamRecord>) template,
                new RecordStringKeyFormatter(loginParamRecordPrefix),
                new MapStructBeanTransformer<>(
                        LoginParamRecord.class, FastJsonLoginParamRecord.class, BeanMapper.class
                )
        );
    }

    @Bean(name = "acckeeper.protectDetailRecordCacheDelegate")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<RecordKey, ProtectDetailRecord, FastJsonProtectDetailRecord>
    protectDetailRecordCacheDelegate() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonProtectDetailRecord>) template,
                new RecordStringKeyFormatter(protectDetailRecordPrefix),
                new MapStructBeanTransformer<>(
                        ProtectDetailRecord.class, FastJsonProtectDetailRecord.class, BeanMapper.class
                )
        );
    }

    @Bean(name = "acckeeper.deriveHistoryCacheDelegate")
    @SuppressWarnings("unchecked")
    public RedisBatchBaseCache<LongIdKey, DeriveHistory, FastJsonDeriveHistory> deriveHistoryCacheDelegate() {
        return new RedisBatchBaseCache<>(
                (RedisTemplate<String, FastJsonDeriveHistory>) template,
                new LongIdStringKeyFormatter(deriveHistoryPrefix),
                new MapStructBeanTransformer<>(DeriveHistory.class, FastJsonDeriveHistory.class, BeanMapper.class)
        );
    }
}
